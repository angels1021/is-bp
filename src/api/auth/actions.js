/*
 * LoginActions
 *
 * redux actions to describe login/logout actions
 *
 * */
import { createAction } from 'redux-actions';
import { createFormAction } from 'redux-form-saga';
import {
  LOGIN_REQUEST,
  LOGOUT_REQUEST,
  AUTH_REQUEST,
  AUTH_ERROR,
  AUTH_SUCCESS,
  AUTH_SET
} from './constants';

/**
 * Tells the app we want to login a user
 * @param  {object} formData          The formData we're sending for login
 * @param  {string} formData.username The username of the user to login
 * @param  {string} formData.password The password of the user to login
 * @param  {string} formData.location The location to navigate to after login
 * @param  {string} formData.code     The card code of the user to login (optional)
 *
 * @return {object} action object - { type: LOGIN_REQUEST, payload: formData{} }
 */
export const login = createAction(LOGIN_REQUEST, (formData) => formData.toObject()); // (formData) => formData

/**
 * Tells the app we want to log out a user
 *
 * @param {string/number} userId - the user to logout;
 *
 * @return {object} action object - { type: LOGOUT_REQUEST, payload: userId }
 */
export const logout = createAction(LOGOUT_REQUEST); // (userId) => userId

/**
 * Sets the 'error' state to error received (both login and logout use the same reducer)
 * @see http://redux-form.com/6.5.0/docs/api/SubmissionError.md/
 * for SubmissionError
 *
 * @param {object} Error/SubmissionError The error received when trying to make a request.
 *
 * @return {object} action object - { type: AUTH_ERROR, payload: error }
 */
export const authError = createAction(AUTH_ERROR); // (SubmissionError) => SubmissionError
// export const authError = (error) => ({
//   type: AUTH_ERROR,
//   payload: error
// }); // (error) => error

/**
 * Sets the 'pending' state to requested action (both login and logout use the same reducer)
 *
 * @param {string} type The action type making the request
 *
 * @return {object} action object - { type: AUTH_REQUEST, payload: LOGIN_REQUEST }
 */
export const authRequest = createAction(AUTH_REQUEST); // (type) => type

/**
 * Sets the 'user' state to requested response (both login and logout use the same reducer)
 *
 * @param {object} [user={}] The new user object to set. on logout it's unnecessary
 * @param {string} user.id The user id
 * @param {string} user.uuid The user's uuid
 * @param {string} user.role The user's authorized role
 * @param {string} user.firstName The user's first name
 * @param {string} user.lastName The user's last name
 *
 * @return {object} action object - { type: AUTH_SUCCESS, payload: user{} }
 */
export const authSuccess = createAction(AUTH_SUCCESS, (user) => user || {});

/**
 * Sets the authentication state of the application
 *
 * @param  {boolean} isLoggedIn
 */
export const authSet = createAction(AUTH_SET, (isLoggedIn) => Boolean(isLoggedIn));

export const loginAction = createFormAction(login, [AUTH_SUCCESS, AUTH_ERROR]);
