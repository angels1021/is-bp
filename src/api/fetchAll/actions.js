/**
 * fetchAll api actions
 */
import { FETCH_ALL } from './constants';
import { createRequestId } from './utils';
/**
 * action creator for a fetchAll request.
 * expects an options object
 *
 * @param {string} requestId           - unique page/module id to track in global async reducer
 * @param {generator} fetchGenerator   - generator to call when ready to fetch resources
 * @param {string} [parentId]          - unique page/module id to wait for before starting calls
 * @param {function} [resolveSelector] - selector function that must resolve to true before starting calls
 */
export const fetchAllCreateSaga = ({ requestId, fetchGenerator, parentId, resolveSelector }) => ({
  type: FETCH_ALL,
  payload: {
    requestId: createRequestId(requestId),
    parentId: createRequestId(parentId),
    fetchGenerator,
    resolveSelector
  }
});

/**
 * actions creator for fetchAll events
 * should be listened to in individual module/page sagas
 *
 * @param moduleId
 */
export const fetchAll = (moduleId) => ({
  type: createRequestId(moduleId),
  payload: { id: moduleId }
});

