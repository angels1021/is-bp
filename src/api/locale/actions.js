/**
 * Locale api actions
 */
import {
  CHANGE_LOCALE_REQUEST,
  CHANGE_LOCALE_SUCCESS,
  CHANGE_LOCALE_FAIL,
  CHANGE_LOCALE_CANCEL
} from './constants';

/**
 * changeLocale - action creator
 *
 * @param {string} locale - two character language code, ie 'en'.
 * @param {object[]} request - request translations for the modules specified - passed to api/translations.
 * @property {string} request[].module - ie 'app'.
 * @property {string} request[].page ie  -'common'.
 * @param {string} moduleId - the caller module who's reducer should catch this actions
 */
export const changeLocale = (locale, request, moduleId) => ({
  type: CHANGE_LOCALE_REQUEST,
  payload: { locale, request, moduleId }
});

/**
 * changeLocaleSuccess - action creator
 *
 * @param {string} locale - two character language code, ie 'en'.
 * @param {string} moduleId - the caller module who's reducer should catch this actions
 */
export const changeLocaleSuccess = (locale, moduleId) => ({
  type: CHANGE_LOCALE_SUCCESS,
  payload: { locale, moduleId }
});

/**
 * changeLocaleFail - action creator
 *
 * @param {object} error - the response error
 * @param {string} moduleId - the caller module who's reducer should catch this actions
 */
export const changeLocaleFail = (error, moduleId) => ({
  type: CHANGE_LOCALE_FAIL,
  payload: { error, moduleId }
});

/**
 * changeLocaleCancel - action creator
 *
 * @param {string} moduleId - the caller module who's reducer should catch this actions
 */
export const changeLocaleCancel = (moduleId) => ({
  type: CHANGE_LOCALE_CANCEL,
  payload: { moduleId }
});

