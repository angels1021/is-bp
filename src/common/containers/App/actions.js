/**
 * Global action creators
 *
 * here will setup action creators to be used by the global reducer
 * and common actions that apply to any component
 */
import { ASYNC_REQUEST, ASYNC_SUCCESS, ASYNC_FAIL } from './constants';

/**
 * action creator for ASYNC_REQUEST
 * called to start tracking an async call on the global object
 *
 * @param  {string} id - a unique request id to track
 */
export const asyncRequest = (id) => ({
  type: ASYNC_REQUEST,
  payload: { id }
});

/**
 * action creator for ASYNC_SUCCESS
 * called to complete tracking an async call and remove from global.loading list
 *
 * @param  {string} id - the unique request id used in asyncRequest
 */
export const asyncSuccess = (id) => ({
  type: ASYNC_SUCCESS,
  payload: { id }
});

/**
 * action creator for ASYNC_FAIL
 * called to complete tracking an async call and set an error in global.errors list
 *
 * @param {string} id    - the unique request id used in asyncRequest
 * @param {object} error - a new Error object to set as the request error
 */
export const asyncFail = (id, error) => ({
  type: ASYNC_FAIL,
  payload: { id, error }
});
