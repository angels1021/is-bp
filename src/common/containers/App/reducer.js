/*
 * AppReducer
 *
 * The reducer takes care of our data. Using actions, we can change our
 * application state.
 * To add a new action, add it to the switch statement in the reducer function
 *
 * Example:
 * case YOUR_ACTION_CONSTANT:
 *   return state.set('yourStateVariable', true);
 */

import { fromJS } from 'immutable';
import { combineReducers } from 'redux-immutable';
import translationsReducer from 'api/translations/reducer';
import { ASYNC_REQUEST, ASYNC_SUCCESS, ASYNC_FAIL } from './constants';

const deleteInList = (list, value) => list.filter((item) => item !== value);

const initialState = fromJS({
  loading: [],
  errors: {}
});

// track async requests across the app
export const asyncReducer = (state = initialState, action) => {
  const { type } = action;
  switch (type) {
    case ASYNC_REQUEST: {
      const { payload: id } = action;
      return state.deleteIn(['errors', id]).mergeIn(['loading'], [id]);
    }
    case ASYNC_SUCCESS: {
      const { payload: id } = action;
      return state.deleteIn(['errors', id]).set('loading', deleteInList(state.get('loading'), id));
    }
    case ASYNC_FAIL: {
      const { id, error } = action.payload;
      return state.setIn(['errors', id], error).set('loading', deleteInList(state.get('loading'), id));
    }
    default:
      return state;
  }
};

export default combineReducers({
  async: asyncReducer,
  translations: translationsReducer
});
