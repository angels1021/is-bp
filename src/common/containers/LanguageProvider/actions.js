/*
 *
 * LanguageProvider actions factory
 * since all pages share the same language structure
 * we can use a factory to create their actions
 *
 */

const LanguageActionsFactory = (
  CHANGE_LOCALE_REQUEST,
  CHANGE_LOCALE_SUCCESS,
  REQUEST_MESSAGES,
  MESSAGES_PENDING,
  MESSAGES_SUCCESS,
  MESSAGES_ERROR
) => ({
  /**
   * changeLocale action creator
   *
   * @param {string} locale - locale identifier to request
   * @param {object} request - locale modules request object
   * @property {string} module - name of the module to fetch
   * @property {string} page - name of the page messages to fetch
   */
  changeLocale: (locale, request) => ({
    type: CHANGE_LOCALE_REQUEST,
    locale,
    request
  }),


  /**
   * changeLocaleSuccess action creator
   *
   * @param {string} locale
   * @param {array} response
   * @property {object} responseItem
   * @property {string} responseItem.module
   * @property {string} responseItem.page
   * @property {object} responseItem.messages
   */
  changeLocaleSuccess: (locale, response) => ({
    type: CHANGE_LOCALE_SUCCESS,
    locale,
    response
  }),
  /**
   * requestMessages action creator
   *
   * @param {string} locale - locale identifier to request
   * @param {object} request - locale modules request object
   * @property {string} module - name of the module to fetch
   * @property {string} page - name of the page messages to fetch
   */
  requestMessages: (locale, request) => ({
    type: REQUEST_MESSAGES,
    locale,
    request
  }),

  /**
   * requestPending action creator
   *
   * set the messages state as pending
   */
  requestPending: () => ({
    type: MESSAGES_PENDING
  }),

  /**
   * requestSuccess action creator
   *
   * @param {string} locale - locale identifier to request
   * @param {array} response
   * @property {object} responseItem
   * @property {string} responseItem.module
   * @property {string} responseItem.page
   * @property {object} responseItem.messages
   */
  requestSuccess: (locale, response) => ({
    type: MESSAGES_SUCCESS,
    locale,
    response
  }),

  /**
   * requestError action creator
   *
   * @param {object} error - new Error('reason')
   */
  requestError: (error) => ({
    type: MESSAGES_ERROR,
    error
  })
});

export default LanguageActionsFactory;
