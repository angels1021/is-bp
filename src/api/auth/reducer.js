/*
 *
 * LoginReducers
 *
 * */
import { fromJS } from 'immutable';
import { isLoggedIn } from 'utils/auth';
import {
  AUTH_REQUEST,
  AUTH_ERROR,
  AUTH_SUCCESS,
  AUTH_SET
} from './constants';

export const initialStateJS = {
  user: {},
  error: false,
  pending: false,
  isLoggedIn: isLoggedIn()
};

const loginInitialState = fromJS(initialStateJS);

const authReducer = (state = loginInitialState, action) => {
  switch (action.type) {
    case AUTH_SUCCESS: {
      const { payload } = action;
      return state
        .set('error', false)
        .set('pending', false)
        .mergeIn(['user'], payload);
    }
    case AUTH_REQUEST: {
      return state
        .set('error', false)
        .set('pending', action.payload);
    }

    case AUTH_ERROR: {
      return state
        .set('pending', false)
        .set('error', action.payload);
    }

    case AUTH_SET: {
      const { payload } = action;
      return state.set('isLoggedIn', payload);
    }

    default:
      return state;
  }
};

export default authReducer;
