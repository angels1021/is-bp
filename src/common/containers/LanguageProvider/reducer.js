/*
 *
 * LanguageProvider locale reducer factory
 * since all pages share the same language structure
 * we can use a factory to create their reducers
 *
 */

import { fromJS } from 'immutable';
import { DEFAULT_LOCALE } from '../App/constants';

export default function LanguageReducerFactory(
  CHANGE_LOCALE_REQUEST,
  CHANGE_LOCALE_SUCCESS,
  CHANGE_LOCALE_FAIL
) {
  const initialState = fromJS({
    locale: DEFAULT_LOCALE,
    pending: false,
    error: false
  });

  return (state = initialState, action) => {
    const { type, payload } = action;
    switch (type) {
      case CHANGE_LOCALE_REQUEST: {
        state.set('error', false);
        return state.set('pending', payload);
      }
      case CHANGE_LOCALE_FAIL: {
        const { error } = payload;
        state.set('error', error);
        return state.set('pending', false);
      }
      case CHANGE_LOCALE_SUCCESS: {
        const { locale } = payload;
        state.set('locale', locale);
        return state.set('pending', false);
      }
      default:
        return state;
    }
  };
}


/* eslint-disable */

/*
plan
auth
reducer:
  locale,
  messages: {
    page1: {},
    page2: {}
    ....
  }

cr
reducer:
  locale,
  messages: {
    ....
  }

ms
reducer:
  locale,
  messages: {
    ...
  }


//globally?
messages
  en
    app
      common
      ....
    auth
      login
      register
      ...
    cr:
      cart(main)
      payment
      ...
    ms
      reports
      settings
      ....
  'he'
    .....

// global selectors
(state) => state.get('messages')
(messages) => messages.get('en')
(locale) => local.get('app')
(local, app, module) => local.get(module) => app.merge(module).flatten().toJS()


// on locale change
moduleReducer
messages

requestChange(locale, [{module, page}, {module, page}])
watcher take
-> take routes/module/LOCALE_CHANGE_REQUEST
-> call requestChange saga
-> put LOCALE_CHANGE_PENDING payload locale
-> check local messages reducer
-> exists -> put routes/module/LOCALE_CHANGE_SUCCESS payload locale
-> missing
call changeLocale saga: pass payload + existing lang[{}] (put REQUEST_MESSAGES_PENDING)
race: messages saga, requestChange constant
messages saga
-> winner messages saga success
(-> put REQUEST_MESSAGES_SUCCESS payload locale, [{module, page, messages}, {module, page, messages}])
-> put LOCALE_CHANGE_PENDING payload false
-> put LOCALE_CHANGE_SUCCESS payload locale
------
-> winner messages fail
(-> put REQUEST_MESSAGES_PENDING payload false
 -> put REQUEST_MESSAGES_FAIL payload - error) in it's saga
-> put LOCALE_CHANGE_PENDING payload false
-> put LOCALE_CHANGE_FAIL payload - error
------
-> winner take requestChange ->
-> put REQUEST_MESSAGES_PENDING false;
-> call requestChange saga

componentWillMount -> requestMessages(locale, [{module, page}, {module, page}])
put routes/module/REQUEST_MESSAGES payload ^

-> check local messages reducer

// messages selector
(state) => state.get('authPage');
(authPage) => authPage.get('messages');

reducer = (state = { locale, messages }, action) => {
  action.type === `routes/${page}/${CHANGE_LOCAL}`
  state.set('locale', action.locale)

  action.type === `routes/${page}/${ADD_MESSAGES}`
  state.mergeIn(['messages'], action.messages);
}


//define messages
auth.messages.js
export default {
  login: {
    id: auth.page1.login,
    defaultMessage: 'login',
    description: 'login button label'
  }
}
dev
export defineMessages(messages);

prod
export Object.keys(messages).reduce((col, key) => {
  col[key] = { id: messages[key].id }
  return col;
}, {});


// saga messages
3 WATCHERS
listen to auth/REQUEST_MESSAGES
listen to cr/REQUEST_MESSAGES
listen to ms/REQUEST_MESSAGES
payload: { locale, module, page }

dispatch auth/MESSAGES_PENDING
...

callMessagesSaga - single, only api
parameters:
(locale, ids, CHANGE_LOCALE) ->
change locale event is specific so needs to be passed in

race {
messages: api call saga
changeLocale: take(CHANGE_LOCALE)
}

if(messages) {
  success return  messages
  fail return false
} else if (changeLocale) {
  continue to CHANGE_LOCALE saga

}

back to watcher dispatch auth/MESSAGES_SUCCESS
....cr, ms
payload { messages }

or dispatch auth/MESSAGES_ERROR
payload { error }


// saga locale
watchers:
take auth/CHANGE_LOCALE
payload { locale, module, page }
put(auth/REQUEST_MESSAGES)

*/
