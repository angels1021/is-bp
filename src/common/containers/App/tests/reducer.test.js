/**
 * Test Global.async reducer
 */
import { asyncRequest, asyncSuccess, asyncFail } from '../actions';
import { asyncReducer } from '../reducer';

describe('Global.async reducer', () => {
  let state;
  const requestId = 'bubbleWrap';
  beforeEach(() => {
    state = asyncReducer(undefined, {});
  });

  it('should return the initial state', () => {
    expect(state).toMatchSnapshot();
  });

  describe('ASYNC_REQUEST', () => {
    it('should clear any error associated with the passed id and add it to the loading list', () => {
      // act
      const result = asyncReducer(state, asyncRequest(requestId));
      // assert
      expect(result.get('errors').has(requestId)).toBe(false);
      expect(result.get('loading').includes(requestId)).toBe(true);
    });
  });

  describe('ASYNC_SUCCESS', () => {
    it('should clear any error associated with the passed id and remove it from the loading list', () => {
      // act
      const result = asyncReducer(state, asyncSuccess(requestId));
      // assert
      expect(result.get('errors').has(requestId)).toBe(false);
      expect(result.get('loading').includes(requestId)).toBe(false);
    });
  });

  describe('ASYNC_FAIL', () => {
    it('should remove the passed id from the loading list and set passed error in errors map', () => {
      // arrange
      const error = new Error('some error');
      // act
      const result = asyncReducer(state, asyncFail(requestId, error));
      // assert
      expect(result.get('loading').includes(requestId)).toBe(false);
      expect(result.get('errors').has(requestId)).toBe(true);
      expect(result.getIn(['errors', requestId])).toEqual(error);
    });
  });
});
