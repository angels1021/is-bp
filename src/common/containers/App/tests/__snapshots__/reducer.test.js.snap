// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Global.async reducer should return the initial state 1`] = `
Object {
  "errors": Object {},
  "loading": Array [],
}
`;
